keps/sig-node/4815-dra-partitionable-devices/README.md

```diff
@@ -0,0 +1,2437 @@
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+<!--
+This section is incredibly important for producing high-quality, user-focused
+documentation such as release notes or a development roadmap. It should be
+possible to collect this information before implementation begins, in order to
+avoid requiring implementors to split their attention between writing release
+notes and implementing the feature itself. KEP editors and SIG Docs
+should help to ensure that the tone and content of the `Summary` section is
+useful for a wide audience.
+
+A good summary is probably at least a paragraph in length.
+
+Both in this section and below, follow the guidelines of the [documentation
+style guide]. In particular, wrap lines to a reasonable length, to make it
+easier for reviewers to cite specific portions, and to minimize diff churn on
+updates.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+One of the original use-cases for Dynamic Resource Allocation (DRA) was the
+ability to dynamically allocate partitions of a full device (in addition to
+the full device itself). Whereas the traditional device plugin API forces users
+to statically partition devices ahead of time, DRA was designed to allow those
+partitions to be created on demand. This, in turn, leads to increased resource
+utilization as the size of each partitioned device can be matched in real-time
+to the workload requesting it.
+
+As DRA has evolved from what we now call "classic" DRA to "structured
+parameters" this ability to dynamically partition devices has been lost.
+This KEP proposes a method to bring this capability back within the framework
+that "structured parameters" provides. Additionally, it provides primitives to
+help represent both full devices and their partitions in a more compact way
+than is possible today.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+One of the primary motivating examples for supporting partitionable devices
```

Comment:
I don't think this should be blocking but we should have other use cases than just NVIDIA here. I have heard that AMD/Intel also have partitionable GPUs. Could we maybe add a beta requirement to get use cases from other vendors? That way, we can make sure that this feature will satisfy those use cases also?

Reply:
I've pinged @byako to provide some text around the Intel use case. We discussed in the past that their use is covered by the proposed APIs, but I agree, it would be good to add at least one more example from a different vendor here.

Reply:
I also want to use this for TPU topologies. Similar to MIG, you can utilize TPUs in specific topologies. If you have 8 TPUs in a node, not all selections of say, 2 TPUs are valid. So Google is another vendor for this.

I also think we can use it to build a compound device driver, that allocates a GPU and NIC as a unit (the driver decides which GPU/NIC go together). This is simpler for the user than using `matchAttributes`.

Reply:
Thank you folks! I don't think this needs to block the KEP.

Reply:
I almost created a PR for this PR introducing other examples, but then it dawned on me that the proposed approach is very flexible, and for SR-IOV in particular, it's too flexible, see the standalone comment in the PR. Basically, some sort of restriction is needed to the subsequent allocations that would only allow same-sized allocations after the the first one was done.

Reply:
Oh, interesting, is this like you can allocate of VF of a certain bandwidth, but once you pick the size all the rest on that PF have to be the same bandwidth?

Reply:
Yes, but not because the HW cannot do heterogeneous setup, instead it is because it's only possible to change the number of the VFs to non-zero value when there are none:
- 0 VFs -> any number up to supported limit
- N VFs -> 0

If there is already at least one VF, and two are needed, the only way is to disable the already enabled once first:
- 1 VF -> 0 VFs
- 0 VFs -> 2 VFs

I'll put more detailed examples outside of this thread.

Reply:
I think it will be good to capture @johnbelamaric blurb about TPUs as another example. (I don't think we need details for the partitioning scheme for that.)